src/plugins/map/MQ2MapCommands.cpp

```diff
@@ -1174,146 +1181,340 @@ static void DrawMapSetting_SingleFilter(MapFilter filter)
+	{
+		if (options[childIndex] == nullptr)
+			break;
+		MapFilter parentFilter = options[childIndex]->RequiresOption;
+		if (parentFilter != MapFilter::Invalid)
+		{
+			MapFilterOption parent = GetMapFilterOption(parentFilter);
+			for (int parentIndex = 0; parentIndex < maxOptionsSize; parentIndex++)
+			{
+				if (options[parentIndex] == nullptr) break;
+				if (std::strcmp(parent.szName, options[parentIndex]->szName) == 0)
+				{
+					// Shift child to be immediately after parent
+					MapFilterOption* childToMove = options[childIndex];
+					if (childIndex < parentIndex)
+					{
+						// Moving child right until after parent
+						for (int i = childIndex; i < parentIndex; i++)
+						{
+							options[i] = options[i + 1];
+						}
+						options[parentIndex--] = childToMove;
+					}
+					else if(childIndex != parentIndex + 1) // If child is already immediately after parent, don't move
+					{
+						// Moving child left until after parent
+						for (int k = childIndex; k > parentIndex + 1; k--)
+						{
+							options[k] = options[k - 1];
+						}
+						options[parentIndex + 1] = childToMove;
+					}
+
+					// Shift child right if next element is also a child and higher alphabetically
+					for (int movedChildIndex = parentIndex + 1;
+						movedChildIndex < maxOptionsSize - 1 && 
+						options[movedChildIndex]->RequiresOption == options[movedChildIndex + 1]->RequiresOption &&
+						std::strcmp(options[movedChildIndex]->szName, options[movedChildIndex + 1]->szName) > 0;
+						movedChildIndex++)
+					{
+						MapFilterOption* temp = options[movedChildIndex + 1];
+						options[movedChildIndex + 1] = options[movedChildIndex];
+						options[movedChildIndex] = temp;
+					}
+
+					// Adjust iterator for movement
+					if (childIndex < parentIndex)
+					{
+						// Current element was moved later and the next element was put in its place, so recheck current.
+						childIndex--;
+					} // else current element was moved earlier in the array, and the element that took its place is one we previously checked, so continue
+
+					continue; // we found the parent, move onto next i value and skip the rest of the j's
+				}
+			}
+		}
+	}
+}
+
+static void BuildOptionArrays()
+{
+	// Build the lists once, cache them for future use
+	if (mapFilterOptions[0] != nullptr)
+		return;
+
+	// Copy MapFilterOptions
+	MapFilterOption* allFilterOptions[maxOptionsSize] = { };
+	for (int i = 0; i < maxOptionsSize; i++)
+	{
+		if (MapFilterOptions[i].ThisFilter != MapFilter::Invalid && MapFilterOptions[i].ThisFilter != MapFilter::All)
+			allFilterOptions[i] = &MapFilterOptions[i];
+	}
+
+	// Sort copied array alphabetically
+	std::sort(allFilterOptions, allFilterOptions + maxOptionsSize,
+		[](MapFilterOption* a, MapFilterOption* b) -> bool
+		{
+			if (a == nullptr || a->szName == nullptr) return false;
+			if (b == nullptr || b->szName == nullptr) return true;
+			for (int i = 0; ; i++)
```

Comment:
we have lots of string utils, use ci_less (mq/include/base/String.h)

Reply:
how do I access ci_less's internal struct nocase_compare?

Reply:
to what end? You can just use ci_equals, ci_find_substr, ci_starts_with, or ci_ends_with

Reply:
oh, I see, this is a container. use a ci_unordered container, and just sort it.

Reply:
ah nvm, I see how it's used now.  I wanted basically this: return ci_less()(a->szName, b->szName); and didn't know the syntax expected.

Reply:
I see you use raw arrays a lot, that's the issue here. Prefer stdlib containers so you don't have to reinvent the wheel on all these kinds of things. I think what you want to do is this:
`std::map<std::string_view, MapFilterOption*, ci_less> allFilterOptions(MapFilterOptions.begin(), MapFilterOptions.end());`

that will give you a map sorted by key using ci_less, in way fewer lines.

Reply:
yes for some reason I felt compelled to not bring in other types when I worked on this ages ago, and instead tried to reuse existing types that were in play for the plugin.  Brainiac suggested vectors earlier in the PR so I've been trying to update everywhere to using these vectors and like you mentioned, it has already condensed considerably.  Your input is very much appreciated, when I get this is a better state after the refactors, please let me know what you think.  It's still fluid as I try to update some really old patterns from the original submission years ago.

Reply:
well, that's actually not right because MapFilterOptions isn't a map, you'd have to declare the map then loop (with a range loop!) MapFilterOptions and insert that way. Or use std::transform.

Reply:
ci_less()(a, b) was the correct answer